### PR TITLE
[PH] Support configuring transaction type to send in trx generator through JSON transaction description

### PR DESCRIPTION
**Pull request by oschwaldp-oci**
### JSON Transaction Description Support
Add the ability to configure the type of transaction to generate/send from a JSON transaction description (file, str, etc.)
Added functionality to transaction generator to take input of abi from file, transaction action type, and action data from a string or file json description and generate transactions matching that description.

#### Transaction Generator now takes additional arguments:

- `action-name` The action name applied to the provided action data input
- `action-data` The path to the json action data file or json action data description string to use
- `abi-file` The path to the contract abi file to use for the supplied transaction action data

#### `launch_transaction_generators.py` also now supports:
-  `action_name`           The action name applied to the provided action data input
-  `action_data`           The path to the json action data file or json action data description string to use
-  `abi_file`              The path to the contract abi file to use for the supplied transaction action data

####  `performance_test_basic.py` now supports:
- `--exercise-trx-specification` Test Transaction Generator: abi, action name, action data api

### Other Changes/Additions:

Rename handler_account to contract_owner_account for more accurate description.

Add integration test exercising the new functionality that builds transactions from abi, action type and action data.  Makes use of the performance_test_basic with new argument '--exercise-trx-specification', but generates the transfer transactions from the abi, action name and action description in the test.

### CODE REVIEW

File: tests/performance_tests/CMakeLists.txt [hunk 1]
```diff
@@ -10,6 +10,7 @@ configure_file(genesis.json genesis.json COPYONLY)
 configure_file(validate_nodeos_plugin_args.py validate_nodeos_plugin_args.py COPYONLY)
 
 add_test(NAME performance_test_basic COMMAND tests/performance_tests/performance_test_basic.py -v -p 1 -n 1 --target-tps 20 --tps-limit-per-generator 10 --clean-run WORKING_DIRECTORY ${CMAKE_BINARY_DIR})
+add_test(NAME performance_test_basic_ex_trx_spec COMMAND tests/performance_tests/performance_test_basic.py -v -p 1 -n 1 --target-tps 20 --tps-limit-per-generator 10 --clean-run --exercise-trx-specification WORKING_DIRECTORY ${CMAKE_BINARY_DIR})
```
Comment (heifner) [inline tests/performance_tests/CMakeLists.txt:13]: I was expecting to see data passed in instead of `--exercise-trx-specification`

File: tests/performance_tests/performance_test_basic.py [hunk 2]
```diff
@@ -275,10 +277,15 @@ def runTpsTest(self) -> PtbTpsTestResult:
 
         self.data.startBlock = self.waitForEmptyBlocks(self.validationNode, self.emptyBlockGoal)
         tpsTrxGensConfig = ltg.TpsTrxGensConfig(targetTps=self.ptbConfig.targetTps, tpsLimitPerGenerator=self.ptbConfig.tpsLimitPerGenerator)
+
+        abiFile = Path("unittests")/"contracts"/"eosio.token"/"eosio.token.abi" if self.ptbConfig.exerciseTrxSpecification else None
+        actionName = "transfer" if self.ptbConfig.exerciseTrxSpecification else None
+        actionData = f'{{"from":"{self.account1Name}","to":"{self.account2Name}","quantity":"0.0001 CUR","memo":"transaction specified"}}'  if self.ptbConfig.exerciseTrxSpecification else None
```
Comment (heifner) [inline tests/performance_tests/performance_test_basic.py:283]: I would have expected all of this to be provided on the command line or in some kind of config file.

File: tests/trx_generator/trx_generator.cpp [hunk 3]
```diff
@@ -202,4 +205,117 @@ namespace eosio::testing {
...
+            << "Stop Generation (form potential ongoing generation in preparation for starting new generation run)."
+            << std::endl;
+      stop_generation();
+
+      std::cout << "Create Initial Transaction with action data." << std::endl;
+      abi_serializer abi = abi_serializer(fc::json::from_file(_abi_data_file_path).as<abi_def>(), abi_serializer::create_yield_function( abi_serializer_max_time ));
+      fc::variant unpacked_action_data_json = json_from_file_or_string(_action_data_file_or_str);
+      std::cout << "action data variant: " << fc::json::to_pretty_string(unpacked_action_data_json) << std::endl;
+
+      bytes packed_action_data_string;
```
Comment (heifner) [inline tests/trx_generator/trx_generator.cpp:245]: I would rename to `packed_action_data` as it is not a `string`.

File: tests/trx_generator/trx_generator.cpp [hunk 4]
```diff
@@ -202,4 +205,117 @@ namespace eosio::testing {
...
+   }
+
+   void trx_generator::stop_generation() {
+      ilog("Stopping transaction generation");
+
+      if(_txcount) {
+         ilog("${d} transactions executed, ${t}us / transaction", ("d", _txcount)("t", _total_us / (double) _txcount));
+         _txcount = _total_us = 0;
+      }
+   }
```
Comment (heifner) [inline tests/trx_generator/trx_generator.cpp:320]: `tear_down`, `push_transaction`, `generate_and_send`, and `stop_generation()` are all copy/paste. Seems like a base class is in order for some of this functionality boilerplate.

**Review by heifner — CHANGES_REQUESTED** [review]

File: tests/trx_generator/trx_generator.cpp [hunk 5]
```diff
@@ -202,4 +205,117 @@ namespace eosio::testing {
...
+         try {
+            return fc::json::from_string(file_or_str, ptype);
+         } EOS_RETHROW_EXCEPTIONS(json_parse_exception, "Fail to parse JSON from string: ${string}", ("string", file_or_str));
+      }
+   }
+
+   trx_generator::trx_generator(std::string chain_id_in, const std::string& abi_data_file, std::string contract_owner_account, std::string auth_account, std::string action_name,
+         const std::string& action_data_file_or_str, int64_t trx_expr, const std::string& private_key_str, std::string lib_id_str, std::string log_dir) :
+      _provider(), _chain_id(chain_id_in), _abi_data_file_path(abi_data_file), _contract_owner_account(contract_owner_account), _auth_account(auth_account), _action(action_name), _action_data_file_or_str(action_data_file_or_str),
+      _trx_expiration(trx_expr*1000000), _private_key(fc::crypto::private_key(private_key_str)),
```
Comment (heifner) [inline tests/trx_generator/trx_generator.cpp:226]: Would like to see `fc::microseconds` used instead of `int64_t`

File: tests/performance_tests/performance_test_basic.py (hunk 2 above)
Comment (oschwaldp-oci) [reply tests/performance_tests/performance_test_basic.py]: Addressed in updates, most recently: [777d9d3](https://github.com/AntelopeIO/leap/pull/626/commits/777d9d30bf5918a4fb6f106d45eef3f69d3b59f4)

File: tests/performance_tests/CMakeLists.txt (hunk 1 above)
Comment (oschwaldp-oci) [reply tests/performance_tests/CMakeLists.txt]: Addressed in updates, most recently: [777d9d3](https://github.com/AntelopeIO/leap/pull/626/commits/777d9d30bf5918a4fb6f106d45eef3f69d3b59f4)

File: tests/trx_generator/trx_generator.cpp (hunk 3 above)
Comment (oschwaldp-oci) [reply tests/trx_generator/trx_generator.cpp]: Addressed: [4817492](https://github.com/AntelopeIO/leap/pull/626/commits/48174929b2ba59cce7c53c7a6291843547e7d521)

File: tests/trx_generator/trx_generator.cpp (hunk 4 above)
Comment (oschwaldp-oci) [reply tests/trx_generator/trx_generator.cpp:268]: Addressed: [03750c9](https://github.com/AntelopeIO/leap/pull/626/commits/03750c994aa61d4877ae708c2920245ff8e9526b)

File: tests/trx_generator/trx_generator.cpp (hunk 5 above)
Comment (oschwaldp-oci) [reply tests/trx_generator/trx_generator.cpp]: Addressed: [03750c9](https://github.com/AntelopeIO/leap/pull/626/commits/03750c994aa61d4877ae708c2920245ff8e9526b)

File: tests/trx_generator/trx_generator.cpp [hunk 6]
```diff
@@ -167,7 +156,69 @@ namespace eosio::testing {
...
+       : trx_generator_base(chain_id_in, contract_owner_account, trx_expr, lib_id_str, log_dir), _abi_data_file_path(abi_data_file), _auth_account(auth_account),
+         _action(action_name), _action_data_file_or_str(action_data_file_or_str), _private_key(fc::crypto::private_key(private_key_str)) {}
+
+   bool trx_generator::setup() {
+      _nonce_prefix = 0;
+      _nonce = static_cast<uint64_t>(fc::time_point::now().sec_since_epoch()) << 32;
+
+      std::cout
+            << "Stop Generation (form potential ongoing generation in preparation for starting new generation run)."
+            << std::endl;
```
Comment (heifner) [inline tests/trx_generator/trx_generator.cpp:185]: This can be a separate PR if you like. Most of trx_generator uses `xlog()` statements instead of `cout`.

**Review by heifner — APPROVED** [review]

File: tests/trx_generator/trx_generator.cpp (hunk 6 above)
Comment (oschwaldp-oci) [reply tests/trx_generator/trx_generator.cpp]: Good call.  Addressed here: [d867824](https://github.com/AntelopeIO/leap/pull/626/commits/d86782461b5914287da0e9b2ac48689013a155c4)

**Review by heifner — APPROVED** [review]
